0.preprocess-sites/4.image-and-segmentation-qc.py

```diff
@@ -504,14 +508,48 @@
         )
 
     image_corr_df = (
-        pd.concat(image_corr_list).drop_duplicates(subset="site_location").reset_index()
+        pd.concat(image_corr_list).drop_duplicates(subset="site").reset_index()
+    )
+
+    image_corr_df["min_corr_value"] = (
+        image_corr_df[corr_qc_cols].astype(float).max(skipna=True, axis=1)
     )
 
     for col in corr_qc_cols:
         image_corr_df.loc[(image_corr_df[col] >= correlation_threshold), col] = "pass"
 
+    for col in corr_qc_cols:
+        if len(image_corr_df[col].unique()) == 1:
+            image_corr_df = image_corr_df.drop(col, axis=1)
+
     if len(image_corr_df.index) > 0:
         corr_output_file = pathlib.Path(results_output, "flagged_correlations.csv")
 
         if check_if_write(corr_output_file, force, throw_warning=True):
             image_corr_df.to_csv(corr_output_file)
+
+# By-well visualization of sites with poor correlations
+if len(corr_qc_cols) > 0:
+    image_corr_df = image_corr_df.merge(loc_df, on="site")
+
+    corr_by_well_gg = (
+        gg.ggplot(image_corr_df, gg.aes(x="x_loc", y="y_loc"))
+        + gg.geom_point(gg.aes(fill="min_corr_value"), size=10)
+        + gg.geom_text(gg.aes(label="site"), color="lightgrey")
+        + gg.facet_wrap("~Metadata_Well")
+        + gg.coord_fixed()
+        + gg.theme_bw()
+        + gg.ggtitle(f"Failed Correlations \n {plate}")
+        + gg.theme(
+            axis_text=gg.element_blank(),
+            axis_title=gg.element_blank(),
+            strip_background=gg.element_rect(colour="black", fill="#fdfff4"),
+        )
+        + gg.labs(fill="Minimum correlation in image set")
```

Comment:
Can we be slightly more specific in the plot title and/or axis labels? Is this visualizing the dapi/hoechst correlation between ISS and cell painting?

Reply:
It's a plate overview so I'm not sure what axis labels would add?
This is looking at the scores for all pairwise image correlations (not just DAPI between BC and CP). It only plots sites that have image correlations that fail to pass the QC threshold and heatmaps the worst pairwise correlation of the image set. It's not a generally super informative or necessary plot, but it would help us visually catch the circumstance where 1) one region of the well is aligning particularly badly or 2) images after a certain point are all aligning badly.

Reply:
That being said, I'm still not sure what a more informative title would be, and open to suggestions!

Reply:
Ah got it. I think I was being thrown off by the title "Failed Correlations". Correlations between what? How about "Failed Image Channel Correlations\n{plate}"? I will add as a suggestion, but feel free to further modify!